FHEM/00_SIGNALduino.pm

```diff
@@ -2187,95 +2187,112 @@ sub SIGNALduino_Parse_MU($$$$@)
-
 			Debug "Searching in patternList: ".Dumper(\%patternList) if($debug);
 			next if (!$valid) ;
 
-	
+			my @msgStartLst;
+			my $startStr="";
+			my $message_start=0 ;
+			my $startLogStr="";
+			
+			if (defined($ProtocolListSIGNALduino{$id}{start}))	# wenn start definiert ist, dann startStr ermitteln und in rawData suchen und in der rawData alles bis zum startStr abschneiden
+			{
+				@msgStartLst = $ProtocolListSIGNALduino{$id}{start};
+				Debug "msgStartLst: ".Dumper(@msgStartLst)  if ($debug);
+				
+				if ( ($startStr=SIGNALduino_PatternExists($hash,@msgStartLst,\%patternList,\$rawData)) eq -1)
+				{
+					SIGNALduino_Log3 $name, 5, "$name: start pattern for MU Protocol id $id -> $ProtocolListSIGNALduino{$id}{name} mismatches, aborting"  ;
+					$valid=0;
+					next;
+				};
+				Debug "startStr is: $startStr" if ($debug);
+				
+				if ($rawData =~ m/$startStr/) {
+					$message_start = $-[0]+ length($startStr);
+					$rawData = substr($rawData, $message_start);
+					$startLogStr = "Start: $startStr cut Pos $message_start" . "; ";
+					Debug "rawData = $rawData" if ($debug);
+					Debug "startStr $startStr found. Message starts at $message_start" if ($debug);
+				} else {
+					Debug "startStr $startStr not found." if ($debug);
+					next;
+				}
+			}
+			
 			my %patternLookupHash=();
 
 			#Debug "phash:".Dumper(%patternLookupHash);
 			my $pstr="";
+			my $zeroStr ="";
+			my $oneStr ="";
+			my $floatStr ="";
+			my $signal_regex;
+			my $start_regex;
+			
 			$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{one}},\%patternList,\$rawData)) >=0;
 			Debug "Found matched one" if ($debug && $valid);
-			my $oneStr=$pstr if ($valid);
-			$patternLookupHash{$pstr}="1" if ($valid); ## Append one to our lookuptable
+			$oneStr=$pstr if ($valid);
+			$patternLookupHash{$pstr}="1" if ($valid); 		## Append one to our lookuptable
 			Debug "added $pstr " if ($debug && $valid);
-
-			my $zeroStr ="";
+			
 			if (scalar @{$ProtocolListSIGNALduino{$id}{zero}} >0) {
 				$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{zero}},\%patternList,\$rawData)) >=0;
 				Debug "Found matched zero" if ($debug && $valid);
-				$zeroStr=$pstr if ($valid);
-				$patternLookupHash{$pstr}="0" if ($valid); ## Append zero to our lookuptable
+				$zeroStr='|' . $pstr if ($valid);
+				$patternLookupHash{$pstr}="0" if ($valid);	## Append zero to our lookuptable
 				Debug "added $pstr " if ($debug && $valid);
 			}
 
-			if (defined($ProtocolListSIGNALduino{$id}{float}))
-			{
+			if (defined($ProtocolListSIGNALduino{$id}{float})) {
 				$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{float}},\%patternList,\$rawData)) >=0;
 				Debug "Found matched float" if ($debug && $valid);
-				$patternLookupHash{$pstr}="F" if ($valid); ## Append float to our lookuptable
+				$floatStr='|' . $pstr if ($valid);
```

Comment:
Ich verstehe, wieso Du das oder mit in die Variable packst, aber genau genommen ist der Inhalt jetzt nicht mehr der floatStr sondern oder$floatstr .

mit `sprintf ("%s|%s|%s", $oneStr,$zeroStr,$floatStr); ` bekommt man den Regex String.
Wenn man alles bis auf ein abschließendes | behalten möchte, macht das folgende Regex:
$signal_regex =~ m/.*[^\|]/

Wenn wir die Variablen $oneStr usw. nicht benötigen, können wir die aber auch ganz weglassen und alles in $signal_regex schreiben. Vielleicht bekomme ich ja noch eine Idee, wenn ich den code weiter durchschaue.

Reply:
Nein, die hat es nur am Ende entfernt. Dass es zero nicht gibt würde ich nicht annehmen, Float kann sein.
Wenn doppelte "oder" entfernt werden sollen geht das nur mit search / replace glaube ich.

Reply:
Die ID 76 sind die Kabellose LED-Weihnachtskerzen von Dir

Reply:
Ah ich erinnere mich. Da wurde an/aus über die Signal Länge gesteuert. :)